File: src/leetcode/python/ValidParentheses_20.py
class ValidParentheses_20(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        res = []
        for c in s:
            if c == '[':
                res.append(']')
            elif c == '{':
                res.append('}')
            elif c == '(':
                res.append(')')
            elif len(res) == 0 or res.pop() != c:
                return False
        return len(res) == 0

File: src/leetcode/java/ValidParentheses_20.java
package leetcode.java;

import java.util.LinkedList;

public class ValidParentheses_20 {
  public boolean isValid(String s) {
    char[] chs = s.toCharArray();
    LinkedList<Character> stack = new LinkedList<>();
    for (char c : chs) {
      if (stack.isEmpty()) {
        stack.addLast(c);
        continue;
      }
      char p = stack.getLast();
      switch (c) {
        case '}': {
          if (p == '{') stack.removeLast();
          else return false;
          break;
        }
        case ']': {
          if (p == '[') stack.removeLast();
          else return false;
          break;
        }
        case ')': {
          if (p == '(') stack.removeLast();
          else return false;
          break;
        }
        default:
          stack.addLast(c);
          break;
      }
    }
    return stack.isEmpty();
  }
}
